include/deal.II/particles/particle_handler.h

```diff
@@ -0,0 +1,480 @@
+                      const MPI_Comm mpi_communicator,
+                      const unsigned int n_properties = 0);
+
+      /**
+       * Clear all particle related data.
+       */
+      void clear();
+
+      /**
+       * Only clear particle data, but keep cache information about number
+       * of particles. This is useful during reorganization of particle data
+       * between processes.
+       */
+      void clear_particles();
+
+      /**
+       * Return an iterator to the first particle.
+       */
+      ParticleHandler<dim,spacedim>::particle_iterator begin() const;
+
+      /**
+       * Return an iterator to the first particle.
+       */
+      particle_iterator begin();
+
+      /**
+       * Return an iterator past the end of the particles.
+       */
+      particle_iterator end() const;
+
+      /**
+       * Return an iterator past the end of the particles.
+       */
+      particle_iterator end();
+
+      /**
+       * Return a pair of particle iterators that mark the begin and end of
+       * the particles in a particular cell. The last iterator is the first
+       * particle that is no longer in the cell.
+       */
+      particle_iterator_range
+      particles_in_cell(const typename parallel::distributed::Triangulation<dim,spacedim>::active_cell_iterator &cell);
+
+
+      /**
+       * Return a pair of particle iterators that mark the begin and end of
+       * the particles in a particular cell. The last iterator is the first
+       * particle that is no longer in the cell.
+       */
+      particle_iterator_range
+      particles_in_cell(const typename parallel::distributed::Triangulation<dim,spacedim>::active_cell_iterator &cell) const;
+
+      /**
+       * Remove a particle pointed to by the iterator.
+       */
+      void
+      remove_particle(const particle_iterator &particle);
+
+      /**
+       * Insert a particle into the collection of particles. Return an iterator
+       * to the new position of the particle. This function involves a copy of
+       * the particle and its properties. Note that this function is of O(N \log N)
+       * complexity for N particles.
+       */
+      particle_iterator
+      insert_particle(const Particle<dim,spacedim> &particle,
+                      const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell);
+
+      /**
+       * Insert a number of particle into the collection of particles.
+       * This function involves a copy of the particles and their properties.
+       * Note that this function is of O(n_existing_particles + n_particles) complexity.
+       */
+      void
+      insert_particles(const std::multimap<types::LevelInd, Particle<dim,spacedim> > &particles);
+
+      /**
+       * This function allows to register three additional functions that are
+       * called every time a particle is transferred to another process
+       * (i.e. during sorting into cells, during ghost particle transfer, or
```

Comment:
i think we need some further documentation in the class description to explain what `ghost particle` is. Most likely it's something owned by another MPI core, but this MPI core needs to know about it.